File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.go
func productExceptSelf(nums []int) []int {
    if len(nums) == 0 {
        return nums
    }

    a, b := 1, 1
    res := make([]int, len(nums))

    for i, v := range nums {
        if v != 0 {
            b *= v
        }

        if v == 0 && a == 0 && i != 0 {
            b *= v
        }
        a *= v
    }

    for i, v := range nums {
        if v == 0 {
            res[i] = b
        } else {
            res[i] = a / v
        }
    }

    return res
}

File: Data Structures & Algorithms/combination-target-sum/submission-0.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        path = []

        def backtracking(start: int, curr_sum: int):
            # goal:
            if curr_sum == target:
                res.append(list(path))
                return

            # constraint:
            if curr_sum > target:
                return

            # choise:
            for i in range(start, len(nums)):
                path.append(nums[i])
                backtracking(i, curr_sum+nums[i])
                path.pop()
        
        backtracking(0, 0)

        return res
